### PR TITLE
ISSUE-274: Report render array output error in Metadata Display preview

### PR DESCRIPTION
**Pull request by aksm**
Start on resolving #274 with refactor to the below to replace deprecated classes. https://github.com/esmero/format_strawberryfield/blob/2e5cc2a6f882cd2490caccc2cbd5b88b58f70c0f/src/Entity/MetadataDisplayEntity.php#L435-L459

### CODE REVIEW

**Comment by DiegoPino** [comment]
@aksm is still still on the works? Need any help or discussing?

**Comment by aksm** [comment]
@DiegoPino I had a question about error reporting that I can dig back up for our call tomorrow.

File: format_strawberryfield.module [hunk 1]
```diff
@@ -364,3 +364,31 @@ function format_strawberryfield_cron() {
...
+ *   error occurred.
+ *
+ * @throws \ErrorException
+ *   Throw ErrorException for all errors passed from trigger_error().
+ *
+ * @see \Drupal\format_strawberryfield\MetadataDisplayForm::ajaxPreview()
+ */
+function _format_strawberryfield_metadata_preview_error_handler($error_level, $message, $filename = NULL, $line = NULL, $context = NULL) {
+    _drupal_error_handler(E_WARNING, $message, $filename, $line, $context);
+    throw new ErrorException($message, $error_level, 0, $filename, $line);
```
Comment (DiegoPino) [inline format_strawberryfield.module:393]: Oh... smart!

File: src/Entity/MetadataDisplayEntity.php [hunk 2]
```diff
@@ -429,29 +432,35 @@ public function getTwigVariablesUsed() {
...
+   *
    * @return array
    *   A list of used $variables by this template.
    */
-  private function getTwigVariableNames(ModuleNode $nodes): array {
+  private function getTwigVariableNames(ModuleNode|Node|BodyNode $nodes): array {
     $variables = [];
     foreach ($nodes as $node) {
-      if ($node instanceof \Twig_Node_Expression_Name) {
+      if ($node instanceof \Twig\Node\Expression\NameExpression) {
```
Comment (DiegoPino) [inline src/Entity/MetadataDisplayEntity.php:447]: @very nice!

File: src/Entity/MetadataDisplayEntity.php [hunk 3]
```diff
@@ -429,29 +432,35 @@ public function getTwigVariablesUsed() {
...
+   *   A Twig Module Nodes object.
+   *
+   * @param \Twig\Node\Node $nodes
+   *   A Twig Module Nodes object.
+   *
    * @return array
    *   A list of used $variables by this template.
    */
-  private function getTwigVariableNames(ModuleNode $nodes): array {
+  private function getTwigVariableNames(ModuleNode|Node|BodyNode $nodes): array {
```
Comment (DiegoPino) [inline src/Entity/MetadataDisplayEntity.php:444]: Ohhh.. this is new to me! So cool. So 3 params in the docs but really a single multi type one in the function argument?

File: src/Form/MetadataDisplayForm.php [hunk 4]
```diff
@@ -178,11 +178,13 @@ public function save(array $form, FormStateInterface $form_state) {
    * AJAX callback.
    */
   public static function ajaxPreview($form, FormStateInterface $form_state) {
+    set_error_handler('_format_strawberryfield_metadata_preview_error_handler');
```
Comment (DiegoPino) [inline src/Form/MetadataDisplayForm.php:181]: Cool

File: src/Form/MetadataDisplayForm.php [hunk 5]
```diff
@@ -354,7 +397,25 @@ public static function ajaxPreview($form, FormStateInterface $form_state) {
...
+          '#type' => 'details',
+          '#open' => FALSE,
+          '#title' => 'Unused JSON keys',
+          'render' => [
+            'table' => $json_table
+          ],
+        ];
       }
+      restore_error_handler();
+      restore_exception_handler();
```
Comment (DiegoPino) [inline src/Form/MetadataDisplayForm.php:418]: Love this!

**Review by DiegoPino — APPROVED** [review]
Looks good. Let's work on the next step together. I have high dreams about an JS/Interactive JS driven JSON picker

**Pull request merged**